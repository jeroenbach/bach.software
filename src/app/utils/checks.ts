type NullOrUndefined = null | undefined;
type NullOrUndefinedOrEmpty = NullOrUndefined | '';

export function isNullOrUndefined<T>(value: T | NullOrUndefined): value is NullOrUndefined {
  return value === null || value === undefined;
}

export function isNullOrUndefinedOrEmpty<T>(value: T | NullOrUndefinedOrEmpty): value is NullOrUndefinedOrEmpty {
  return isNullOrUndefined(value) || value === '';
}

export function isNotNullOrUndefined<T>(value: T | NullOrUndefined): value is T {
  return !isNullOrUndefined(value);
}

export function isNotNullOrUndefinedOrEmpty<T>(value: T | NullOrUndefinedOrEmpty): value is T {
  return !isNullOrUndefinedOrEmpty(value);
}

export function isFalseOrUndefined<T>(value: T | undefined | false): value is undefined | false {
  return value === false || value === undefined;
}

export function isNumber(value: unknown): value is number {
  return typeof value === 'number';
}
