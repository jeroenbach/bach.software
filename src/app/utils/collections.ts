/**
 * Converts an array of items into a dictionary (object) where the keys are generated
 * by a provided function and the values are the items themselves.
 *
 * @template T - The type of items in the array.
 * @template K - The type of keys in the resulting dictionary.
 *
 * @param {T[]} array - The array of items to convert into a dictionary.
 * @param {(item: T) => K} getKey - A function that takes an item and returns a key for that item.
 *
 * @returns {Record<K, T>} - A dictionary where the keys are generated by the getKey function
 * and the values are the items from the array.
 */
export const toDictionary = <T, K extends PropertyKey>(
  array: T[],
  getKey: (item: T) => K,
): Record<K, T> => {
  return array.reduce(
    (acc, item) => {
      acc[getKey(item)] = item;
      return acc;
    },
    {} as Record<K, T>,
  );
};
