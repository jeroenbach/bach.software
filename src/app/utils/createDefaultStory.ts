/**
 * Helper to create a defaultStory function with intellisense.
 * This way typescript helps us with the parameters & args, which can contain an object tree.
 *
 * Please use as followed: createDefaultStory((_: Story) => ({ decoraters: ... your actual story config })
 * Where the type Story is your StoryObj with metadata. In the end we don't use the parameter, only for type inference.
 */
export const createDefaultStory =
  <StoryMeta, T extends StoryMeta>(
    storyfn: (parameterOnlyUsedToDetermineStoryType: StoryMeta) => T,
  ) =>
  (changeDefaultfn: (story: T) => void = () => {}) => {
    const story = storyfn(null as StoryMeta);
    changeDefaultfn(story);
    return story;
  };
