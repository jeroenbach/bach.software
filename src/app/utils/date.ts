import { useDayjs } from '#dayjs';

/**
 * Format date to default format of MMM D, YYYY
 * @param date The date to format
 * @returns The formatted date string
 */
export function formatDate(date?: string | Date) {
  if (!date)
    return '';

  const dayjs = useDayjs();
  return dayjs(date).format('MMM D, YYYY');
}

export function toDateWithTimeZone(date: string | Date) {
  const dayjs = useDayjs();
  return dayjs(date).format();
}
