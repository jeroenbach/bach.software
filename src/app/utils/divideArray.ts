/**
 * Just like filter it filters an array, but it will also return an array with filtered out values.
 * @param array The Array to divide
 * @param divideFn the divide function
 */
export function divideArray<T>(
  array: T[],
  divideFn: (item: T) => boolean,
): [T[], T[]];
/**
 * Just like filter it filters an array, but it will also return an array with filtered out values.
 * @param array The Array to divide
 * @param divideFn the divide function
 * @param mapFn An optional map function, to make use of the 1x iteration
 */
export function divideArray<T, U>(
  array: T[],
  divideFn: (item: T) => boolean,
  mapFn: (item: T) => U,
): [U[], U[]];

export function divideArray<T, U = T>(
  array: T[],
  divideFn: (item: T) => boolean,
  mapFn?: (item: T) => U,
): [U[], U[]] {
  if (!array || !divideFn)
    return [[], []];

  const matched: U[] = [];
  const unmatched: U[] = [];

  for (const item of array) {
    const mapped = mapFn ? mapFn(item) : (item as unknown as U);
    if (divideFn(item)) {
      matched.push(mapped);
    }
    else {
      unmatched.push(mapped);
    }
  }

  return [matched, unmatched];
}
