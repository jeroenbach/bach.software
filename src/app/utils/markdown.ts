import fm from 'front-matter';
import { marked } from 'marked';

export function readMarkdown<TAttributes>(markdown: string) {
  const { attributes, body } = fm(markdown) as {
    attributes: TAttributes
    body: string
  };
  const bodyHtml = marked(body);
  return {
    attributes,
    body,
    bodyHtml,
  };
}
