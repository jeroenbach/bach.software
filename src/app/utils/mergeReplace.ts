/**
 * Merge two arrays of objects:
 * - If a key in newArr exists in oldArr → replace the old object.
 * - If a key in newArr does not exist in oldArr → append it.
 *
 * @param array - the original array
 * @param itemsToAdd - the replacement array
 * @param keyFn  - function to extract the unique key for each object
 */
export function mergeReplace<T, K>(
  array: readonly T[],
  itemsToAdd: readonly T[],
  keyFn: (item: T) => K,
): T[] {
  const newMap = new Map(itemsToAdd.map(item => [keyFn(item), item]));

  const result: T[] = [];

  // 1. Keep old order, replacing where applicable
  for (const oldItem of array) {
    const key = keyFn(oldItem);
    if (newMap.has(key)) {
      result.push(newMap.get(key)!);
      newMap.delete(key); // remove so we don’t append it later
    }
    else {
      result.push(oldItem);
    }
  }

  // 2. Append leftovers (new items not in oldArr)
  for (const leftover of newMap.values()) {
    result.push(leftover);
  }

  return result;
}
