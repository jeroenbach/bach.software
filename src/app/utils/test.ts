import type { StoryObj } from '@storybook/vue3-vite';
import type { DOMWrapper, VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';

/**
 * Helper to mount a storybook story
 * @param story the storybook story to mount
 * @param argsOverride specify args that you want to override
 * @returns wrapper with mounted component
 */
export function mountStory<Story extends StoryObj>(story: Story, argsOverride?: Story['args']) {
  return mount(story.render, { props: { ...story.args, ...argsOverride } });
}

export async function setInputValue(wrapper: VueWrapper, cssQuery: string, value: string) {
  const input = await find(wrapper, cssQuery);
  await input.setValue(value);
}

export async function getComponent(wrapper: VueWrapper, component: any, cssQuery?: string): Promise<VueWrapper> {
  if (!cssQuery) {
    return wrapper.findComponent(component);
  }

  const specific = await find(wrapper, cssQuery);
  return specific.findComponent(component);
}

export async function getComponentVM(wrapper: VueWrapper, component: any, cssQuery?: string) {
  const fieldComponent = await getComponent(wrapper, component, cssQuery);
  return fieldComponent.vm as any;
}

export async function find(wrapper: VueWrapper, cssQuery: string, havePatience: boolean = true, count = 1): Promise<DOMWrapper<Element>> {
  const element = wrapper.find(cssQuery);

  if (element?.exists() || count > 10 || !havePatience)
    return element;

  await nextTick();
  return await find(wrapper, cssQuery, havePatience, ++count);
}

/**
 * Gets the text of an field element
 */
export async function text(wrapper: VueWrapper, cssQuery: string, havePatience: boolean = true) {
  const element = await find(wrapper, cssQuery, havePatience);
  if (!element.exists())
    return null;
  return element.text();
}

export async function exists(wrapper: VueWrapper, cssQuery: string, havePatience: boolean = true) {
  const element = await find(wrapper, cssQuery, havePatience);
  return element.exists();
}

export async function trigger(wrapper: VueWrapper, cssQuery: string, eventString: string, havePatience: boolean = true) {
  const element = await find(wrapper, cssQuery, havePatience);
  if (!element.exists())
    return;
  await element.trigger(eventString);
}

export async function click(wrapper: VueWrapper, cssQuery: string, times: number = 1, havePatience: boolean = true) {
  for (let i = 0; i < times; i++) {
    await trigger(wrapper, cssQuery, 'click', havePatience);
  }
}
