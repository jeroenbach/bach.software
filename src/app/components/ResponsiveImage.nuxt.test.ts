import { describe, expect, it } from "vitest";
import * as stories from "./ResponsiveImage.stories";
import ResponsiveImage from "./ResponsiveImage.vue";
import { find, mountStory } from "~/utils/test";

describe("ResponsiveImage", () => {
  it("should have the same html output", async () => {
    const w = mountStory(stories.Default);
    expect(w.html()).toMatchSnapshot();
  });

  it("should render alt", async () => {
    const w = mountStory(stories.Default);
    const img = await find(w, "img");
    expect(img.attributes("alt")).toBe("alt text");
  });

  it("should render caption", async () => {
    const w = mountStory(stories.Default);
    const caption = await find(w, "figcaption");
    expect(caption.text()).toBe("Figure caption");
  });

  it("should add correct classes to image", async () => {
    const w = mountStory(stories.Default);
    const img = await find(w, "img");
    expect(img.attributes("class")).toBe(
      "w-full aspect-1/1 bg-slate-50 dark:bg-slate-900 object-cover",
    );
  });

  it("should render a placeholder when no image is provided", async () => {
    const w = mountStory(stories.NoImage);
    const placeholder = await find(w, "figure div");
    expect(placeholder.exists()).toBe(true);
    expect(placeholder.attributes("class")).toBe(
      "w-full aspect-1/1 bg-slate-50 dark:bg-slate-900 object-cover",
    );
  });

  it("should render avif & webp", async () => {
    const w = mountStory(stories.Default);
    const sourceAvif = await find(w, "source[type='image/avif']");
    const sourceWebp = await find(w, "source[type='image/webp']");
    const img = await find(w, "img");
    expect(sourceAvif.exists()).toBe(true);
    expect(sourceWebp.exists()).toBe(true);
    expect(img.exists()).toBe(true);
  });

  it("should set correct sizes with default screen sizes", async () => {
    const w = mountStory(stories.Default);
    const img = await find(w, "img");
    expect(img.attributes("sizes")).toBe(
      "(max-width: 640px) 320px, (max-width: 768px) 640px, (max-width: 1024px) 768px, (max-width: 1280px) 1024px, (max-width: 1536px) 1280px, 1536px",
    );
  });

  it("should add only sizes if its specified", async () => {
    const w = mountStory(stories.OneSize);
    const img = await find(w, "img");
    expect(img.attributes("sizes")).toBe("320px");
  });

  it("should calculate the correct ratios for each screen size", async () => {
    const w = mountStory(stories.DifferentScreenSizesAndRatios);
    const img = await find(w, "img");
    expect(img.classes()).toEqual(
      expect.arrayContaining([
        "xs:aspect-1/1",
        "sm:aspect-2/1",
        "md:aspect-16/9",
        "lg:aspect-1/1",
        "xl:aspect-2/1",
        "xxl:aspect-16/9",
      ]),
    );
  });

  it("should calculate the correct sizes for each screen size", async () => {
    const w = mountStory(stories.DifferentScreenSizesAndRatios);
    const img = await find(w, "img");
    expect(img.attributes("sizes")).toEqual(
      [
        "(max-width: 640px) 640px",
        "(max-width: 768px) 384px",
        "(max-width: 1024px) 341px",
        "(max-width: 1280px) 320px",
        "(max-width: 1536px) 307px",
        "341px",
      ].join(", "),
    );
  });

  it("should calculate only the sizes that are specified", async () => {
    const getImgSizes = async (override?: stories.Story["args"]) =>
      (
        (
          await getComponent(
            mountStory(stories.DifferentScreenSizesAndRatios, override),
            ResponsiveImage,
          )
        ).vm as any
      ).imgSizes;

    const sizes = (remove?: number) =>
      ["xs:640px", "sm:384px", "md:341px", "lg:320px", "xl:307px", "xxl:341px"]
        .filter((_, i) => i !== remove)
        .join(" ");

    expect(await getImgSizes({ partOfScreenExtraSmall: undefined })).toBe(
      sizes(0),
    );
    expect(await getImgSizes({ partOfScreenSmall: undefined })).toBe(sizes(1));
    expect(await getImgSizes({ partOfScreenMedium: undefined })).toBe(sizes(2));
    expect(await getImgSizes({ partOfScreenLarge: undefined })).toBe(sizes(3));
    expect(await getImgSizes({ partOfScreenExtraLarge: undefined })).toBe(
      sizes(4),
    );
    expect(await getImgSizes({ partOfScreen2ExtraLarge: undefined })).toBe(
      sizes(5),
    );
  });

  it("should calculate the correct image sizes", async () => {
    const w = mountStory(stories.DifferentScreensSameSizesAndRatios);
    const img = await find(w, "img");
    expect(img.attributes("sizes")).toBe(
      [
        "(max-width: 640px) 640px",
        "(max-width: 768px) 768px",
        "(max-width: 1024px) 1024px",
        "(max-width: 1280px) 1280px",
        "(max-width: 1536px) 1536px",
        "2048px",
      ].join(", "),
    );

    const w2 = mountStory(stories.DifferentScreensSameSizesAndRatios, {
      partOfScreen: 1 / 2,
    });
    const img2 = await find(w2, "img");
    expect(img2.attributes("sizes")).toBe(
      [
        "(max-width: 640px) 320px",
        "(max-width: 768px) 384px",
        "(max-width: 1024px) 512px",
        "(max-width: 1280px) 640px",
        "(max-width: 1536px) 768px",
        "1024px",
      ].join(", "),
    );
  });
});
