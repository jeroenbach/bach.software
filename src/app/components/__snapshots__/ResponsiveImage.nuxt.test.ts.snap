// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`responsiveImage > should have the same html output 1`] = `
"<figure itemscope="" itemtype="https://schema.org/ImageObject" itemprop="image" class="w-full">
  <picture>
    <source type="image/avif" sizes="(max-width: 640px) 320px, (max-width: 768px) 640px, (max-width: 1024px) 768px, (max-width: 1280px) 1024px, (max-width: 1536px) 1280px, 1536px" srcset="/_ipx/w_320&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 320w, /_ipx/w_640&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 640w, /_ipx/w_768&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 768w, /_ipx/w_1024&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 1024w, /_ipx/w_1280&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 1280w, /_ipx/w_1536&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 1536w, /_ipx/w_2048&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 2048w, /_ipx/w_2560&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 2560w, /_ipx/w_3072&amp;f_avif&amp;q_80/JEROEN-4238-HD.jpeg 3072w">
    <source type="image/webp" sizes="(max-width: 640px) 320px, (max-width: 768px) 640px, (max-width: 1024px) 768px, (max-width: 1280px) 1024px, (max-width: 1536px) 1280px, 1536px" srcset="/_ipx/w_320&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 320w, /_ipx/w_640&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 640w, /_ipx/w_768&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 768w, /_ipx/w_1024&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 1024w, /_ipx/w_1280&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 1280w, /_ipx/w_1536&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 1536w, /_ipx/w_2048&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 2048w, /_ipx/w_2560&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 2560w, /_ipx/w_3072&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 3072w">
    <source type="image/jpg" sizes="(max-width: 640px) 320px, (max-width: 768px) 640px, (max-width: 1024px) 768px, (max-width: 1280px) 1024px, (max-width: 1536px) 1280px, 1536px" srcset="/_ipx/w_320&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 320w, /_ipx/w_640&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 640w, /_ipx/w_768&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 768w, /_ipx/w_1024&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 1024w, /_ipx/w_1280&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 1280w, /_ipx/w_1536&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 1536w, /_ipx/w_2048&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 2048w, /_ipx/w_2560&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 2560w, /_ipx/w_3072&amp;f_jpg&amp;q_80/JEROEN-4238-HD.jpeg 3072w">
    <source type="image/webp" sizes="(max-width: 640px) 320px, (max-width: 768px) 640px, (max-width: 1024px) 768px, (max-width: 1280px) 1024px, (max-width: 1536px) 1280px, 1536px" srcset="/_ipx/w_320&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 320w, /_ipx/w_640&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 640w, /_ipx/w_768&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 768w, /_ipx/w_1024&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 1024w, /_ipx/w_1280&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 1280w, /_ipx/w_1536&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 1536w, /_ipx/w_2048&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 2048w, /_ipx/w_2560&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 2560w, /_ipx/w_3072&amp;f_webp&amp;q_80/JEROEN-4238-HD.jpeg 3072w"><img alt="alt text" itemprop="thumbnailUrl" aria-describedby="v-0_figcaption" class="w-full aspect-1/1 bg-slate-50 dark:bg-slate-900 object-cover" data-nuxt-pic="" src="/_ipx/w_3072&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg" sizes="(max-width: 640px) 320px, (max-width: 768px) 640px, (max-width: 1024px) 768px, (max-width: 1280px) 1024px, (max-width: 1536px) 1280px, 1536px" srcset="/_ipx/w_320&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 320w, /_ipx/w_640&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 640w, /_ipx/w_768&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 768w, /_ipx/w_1024&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 1024w, /_ipx/w_1280&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 1280w, /_ipx/w_1536&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 1536w, /_ipx/w_2048&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 2048w, /_ipx/w_2560&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 2560w, /_ipx/w_3072&amp;f_jpeg&amp;q_80/JEROEN-4238-HD.jpeg 3072w">
  </picture>
  <!--v-if-->
  <figcaption id="v-0_figcaption">Figure caption</figcaption>
</figure>"
`;
